pysph/cpy/jit.py

```diff
@@ -0,0 +1,523 @@
+from functools import wraps
+from textwrap import dedent, wrap
+
+from mako.template import Template
+import numpy as np
+import inspect
+import ast
+import importlib
+import warnings
+import sys
+
+from .config import get_config
+from .cython_generator import get_parallel_range, CythonGenerator
+from .transpiler import (
+    Transpiler,
+    convert_to_float_if_needed,
+    filter_calls,
+    get_external_symbols_and_calls,
+    BUILTINS)
+from .types import (dtype_to_ctype, get_declare_info,
+                    dtype_to_knowntype)
+from .extern import Extern
+from .ast_utils import get_unknown_names_and_calls
+
+import pysph.cpy.array as array
+import pysph.cpy.parallel as parallel
+import pysph.cpy.types as types
+
+
+def annotate(func):
+    def wrapper(func):
+        func.is_jit = True
+        return func
+
+    if func is None:
+        return wrapper
+    else:
+        return wrapper(func)
+
+
+def memoize(f):
+    @wraps(f)
+    def wrapper(obj, *args, **kwargs):
+        if not hasattr(obj, 'cache'):
+            obj.cache = dict()
+        key = tuple([get_ctype_from_arg(arg) for arg in args])
+        if key not in obj.cache:
+            obj.cache[key] = f(obj, *args, **kwargs)
+        return obj.cache[key]
+    return wrapper
+
+
+def getargspec(f):
```

Comment:
Another way to do this instead is:
`getargspec = getattr(inspect, 'getfullargspec', getattr(inspect, 'getargspec'))`